File: app/src/main/java/de/westnordost/streetcomplete/quests/car_wash_type/AddCarWashType.java
package de.westnordost.streetcomplete.quests.car_wash_type;

import android.os.Bundle;
import android.support.annotation.NonNull;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;

public class AddCarWashType extends SimpleOverpassQuestType
{
	@Inject public AddCarWashType(OverpassMapDataDao overpassServer)
	{
		super(overpassServer);
	}

	@Override
	protected String getTagFilters()
	{
		return "nodes, ways with amenity=car_wash and !automated and !self_service";
	}

	public AbstractQuestAnswerFragment createForm()
	{
		return new AddCarWashTypeForm();
	}

	public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
	{
		List<String> values = answer.getStringArrayList(AddCarWashTypeForm.OSM_VALUES);
		if(values != null)
		{
			boolean isAutomated = values.contains(AddCarWashTypeForm.AUTOMATED);
			boolean isSelfService = values.contains(AddCarWashTypeForm.SELF_SERVICE);
			boolean isStaffService = values.contains(AddCarWashTypeForm.SERVICE);

			changes.add("automated", isAutomated ? "yes" : "no");
			changes.add("self_service", isSelfService && !isStaffService ? "yes" : "no");
		}
	}

	@Override public String getCommitMessage()
	{
		return "Add car wash type";
	}

	@Override public int getIcon() { return R.drawable.ic_quest_car_wash; }

	@Override public int getTitle(@NonNull Map<String, String> tags)
	{
		return R.string.quest_carWashType_title;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/car_wash_type/AddCarWashType.kt
package de.westnordost.streetcomplete.quests.car_wash_type

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
import de.westnordost.streetcomplete.data.osm.osmquests.Tags
import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.CAR
import de.westnordost.streetcomplete.ktx.toYesNo
import de.westnordost.streetcomplete.quests.car_wash_type.CarWashType.AUTOMATED
import de.westnordost.streetcomplete.quests.car_wash_type.CarWashType.SELF_SERVICE

class AddCarWashType : OsmFilterQuestType<List<CarWashType>>() {

    override val elementFilter = "nodes, ways with amenity = car_wash and !automated and !self_service"
    override val changesetComment = "Add car wash type"
    override val wikiLink = "Tag:amenity=car_wash"
    override val icon = R.drawable.ic_quest_car_wash

    override val questTypeAchievements = listOf(CAR)

    override fun getTitle(tags: Map<String, String>) = R.string.quest_carWashType_title

    override fun createForm() = AddCarWashTypeForm()

    override fun applyAnswerTo(answer: List<CarWashType>, tags: Tags, timestampEdited: Long) {
        val isAutomated = answer.contains(AUTOMATED)
        tags["automated"] = isAutomated.toYesNo()

        val hasSelfService = answer.contains(SELF_SERVICE)
        val selfService = when {
            hasSelfService && answer.size == 1 -> "only"
            hasSelfService -> "yes"
            else -> "no"
        }
        tags["self_service"] = selfService
    }
}
